python/arcticdb/vector_db/vector_db.py

```diff
@@ -0,0 +1,521 @@
+        namespace
+            The namespace to which `vectors` should be upserted.
+        vectors
+            In the case of a mapping, we expect a mapping whose keys are strings
+            (taken as identifiers of vectors), and whose values are in turn mapping
+            minimally containing a key-value pair 'value' pointing to something that
+            yields a one-dimensional `np.ndarray` of numeric types corresponding to a
+            vector. Each vector must have the same dimensionality.
+
+            In the case of a pandas DataFrame, we expect string columns and the entries to
+            all be of a numeric type.
+
+            In the case of an ndarray, we expect a list of `identifiers` corresponding to
+            the number of vectors. We also expect an `np.ndarray` of `np.ndarray`s
+            (i.e. two dimensions) populated by numeric types. Note that upserting an
+            `np.ndarray` corresponds to upserting the transpose of the equivalent
+            `pd.DataFrame`. For example:
+
+            >>> vdb.upsert("vdb", np.array([[0,1],[2,3]]), identifiers=["a", "b"])
+
+            inserts vectors <0,1> and <2,3>, but
+
+            >>> vdb.upsert("vdb", pd.DataFrame(np.array([0,1],[2,3])))
+
+            inserts vectors <0,2> and <1,3>.
+            identifiers
+        identifiers
+            In the case of an `np.ndarray`, we require a separate collection of identifiers of the vectors.
+        metadata
+            Presently ignored.
+        """
+        symbol_name = f"{namespace}{VECTOR_DB_DISTINGUISHED_SUFFIX}"
+        if metadata:
+            warnings.warn("Metadata are presently ignored.")
+
+        try:  # where symbol_name in self._lib
+            expected_dimensions = self._lib.read_metadata(symbol_name).metadata["dimensions"]
+        except NoDataFoundException:  # symbol_name not in self._lib
+            expected_dimensions = None
+
+        data_frame_to_upsert = _generate_dataframe(vectors, expected_dimensions, identifiers, metadata)
+        actual_dimensions = data_frame_to_upsert.shape[0]
+        if actual_dimensions > self._max_vector_size:
+            raise ValueError(
+                f"VectorDB {self} has maximum vector size {self._max_vector_size}, but you attempted to "
+                f"upsert vectors of size {data_frame_to_upsert.shape[0]}. To store vectors with more than "
+                f"{self._max_vector_size} vectors, you will have to initialise a new VectorDB on a "
+                "fresh library like so:\n\n"
+                ">>> arctic_client.create_library('nomen', LibraryOptions(rows_per_segment = k)\n\n"
+                "where k is the maximum vector size."
+            )
+        if data_frame_to_upsert.isnull().values.any():
+            raise ValueError("Vectors upserted must not contain NaNs.")
+        maximum_component = ((0.5 * np.finfo(np.float64).max) ** 0.5) / actual_dimensions
+        if (abs(data_frame_to_upsert).values > maximum_component).any():
```

Comment:
The `isnull()` check and this one are likely to be quite expensive. Not for now, but worth bearing in mind when profiling.

Reply:
Ok